design/one-pager-consuming-k8s-clusters.md

```diff
@@ -0,0 +1,317 @@
+on-prem clusters) can now be targeted for provisioning workloads. To enable this
+functionality, a user would manually create a `Secret` containing the
+`kubeconfig` information for their cluster in the `my-app-team` namespace, then
+create a `KubernetesTarget` that references it.
+
+### KubernetesApplication Scheduling
+
+As mentioned above, `KubernetesApplication` resources are currently scheduled to
+`KubernetesCluster` claims in their namespace. With the introduction of the
+`KubernetesTarget`, `KubernetesApplication` resources can still be scheduled via
+label selectors, but will now be scheduled to a `KubernetesTarget` in their
+namespace, then use its `connectionSecretRef` to get the `Secret` that was
+propagated to the namespace from the Kubernetes cluster managed resource or
+already existed due to manual creation. Importantly, for a
+`KubernetesApplication` to be able to be scheduled to a Kubernetes cluster, a
+`KubernetesTarget` that points to that cluster managed resource or local
+`Secret` *must* be present in the namespace of the `KubernetesApplication`.
+
+This change isolates the actions of *provisioning* (`KubernetesCluster` or
+cloud-specific managed resource), *publishing* (`KubernetesTarget`), and
+*consuming* (`KubernetesApplication`) Kubernetes clusters, actions which may be
+executed by [almost disjoint] sets of users. Having these actions represented by
+separate custom resources allows for more granular [RBAC] permissions around
+Kubernetes cluster usage in Crossplane.
+
+### Automatic KubernetesTarget Creation Controller
+
+Because the ability to *publish* a Kubernetes cluster provides immense power in
+a Crossplane environment, it is desirable for Kubernetes cluster managed
+resources to be able to be consumed in the namespace that they are claimed
+without a user requiring the ability to create a `KubernetesTarget`. For
+instance, if an infrastructure owner wants to allow for an application team to
+provision some clusters dynamically by creating a `KubernetesCluster` claim in
+their namespace that references a `GKEClusterClass`, that application team
+should be able to then schedule `KubernetesApplication` resources to that
+cluster without requiring the permissions to create a `KubernetesTarget` (which
+would allow them to enable to consumption of any Kubernetes cluster managed
+resource) or asking the infrastructure owner to cerate a `KubernetesTarget` for
+them (which would break the self-service model).
+
+To enable this functionality, a single controller should be implemented in core
+Crossplane that watches for the creation of `KubernetesCluster` claim resources
+and automatically creates a `KubernetesTarget` with the `connectionSecretRef`
+set to the connection `Secret` that was propagated to the namespace.
+Importantly, this controller should also clean up the `KubernetesTarget` when
+the `KubernetesCluster` (and corresponding) `Secret` are deleted. This can be
```

Comment:
How do we feel about the this automatic clean-up? I am accomplishing this by setting an `ownerReference` (see technical implementation section). My thought process here is that we want the life-cycle of automatically created `KubernetesTarget` resources strictly tied to that of the `KubernetesCluster` claim (i.e. if Crossplane creates something it should also clean it up). Currently, I believe all of our managed k8s cluster resource controllers use the [default secret propagation](https://github.com/crossplaneio/crossplane-runtime/blob/0f37bea5496e8dd4a47dd731a83bc42b97a43878/pkg/resource/publisher.go#L62) mechanism, so the `Secret` that the `KubernetesTarget` referenced would be cleaned up with the `KubernetesCluster` claim, rendering the `KubernetesTarget` unusable anyway.

Reply:
This seems reasonable to me.

It's possible someone could delete the `KubernetesCluster` with `--cascade=false` and thus the `KubernetesTarget` becomes orphaned, but I don't think that's a huge issue given that its technically the person doing the deleting asked for, even if it doesn't make sense.